### PR TITLE
Better create entity response handling

### PR DESCRIPTION
**Pull request by samiwh**
#### Description
Improved our handling of create entity response ops to give slightly better logs.

#### Release note
Improved error messages relating to creating entities.

#### Tests
Testing this properly requires more framework.
Will test this in a deployment known to have issues with creating entities timing out.

#### Documentation
Release note.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 1]
```diff
@@ -1523,13 +1521,24 @@ void USpatialReceiver::OnReserveEntityIdsResponse(const Worker_ReserveEntityIdsR
...
+		break;
+	case WORKER_STATUS_CODE_TIMEOUT:
+		UE_LOG(LogSpatialReceiver, Log, TEXT("Create entity request timed out and will not be retried. "
+			"Request id: %d, entity id: %lld, message: %s"), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
+		break;
+	case WORKER_STATUS_CODE_APPLICATION_ERROR:
+		UE_LOG(LogSpatialReceiver, Log, TEXT("Create entity request failed. Either the reservation expired or the entity already existed. "
+			"Request id: %d, entity id: %lld, message: %s"), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
+		break;
+	default: ;
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
	default:
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 2]
```diff
@@ -1523,13 +1521,24 @@ void USpatialReceiver::OnReserveEntityIdsResponse(const Worker_ReserveEntityIdsR
...
-	{
-		UE_LOG(LogSpatialReceiver, Verbose, TEXT("Create entity request succeeded: request id: %d, entity id: %lld, message: %s"), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
+	switch (static_cast<Worker_StatusCode>(Op.status_code))
+	{
+	case WORKER_STATUS_CODE_SUCCESS:
+		UE_LOG(LogSpatialReceiver, Verbose, TEXT("Create entity request succeeded. "
+			"Request id: %d, entity id: %lld, message: %s"), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
+		break;
+	case WORKER_STATUS_CODE_TIMEOUT:
+		UE_LOG(LogSpatialReceiver, Log, TEXT("Create entity request timed out and will not be retried. "
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Didn't we decided that this error message is incorrect?

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: We did. Will push the fix. Was looking at whether this could actually be improved further, but it probably isn't worth the time right now.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 1 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Ah yep. Will fix.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -1586,18 +1584,33 @@ void USpatialReceiver::OnReserveEntityIdsResponse(const Worker_ReserveEntityIdsR
...
+			"Request id: %d, entity id: %lld, message: %s"), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
+		break;
 	}
 
 	if (CreateEntityDelegate* Delegate = CreateEntityDelegates.Find(Op.request_id))
 	{
-		Delegate->ExecuteIfBound(Op);
+		if (Delegate->ExecuteIfBound(Op))
+		{
+			return;
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Is it intentional this `return` is here? (since it doesn't seem to be related to logging)

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Without it we'd be logging that there is no actor channel whenever this message comes through for something that isn't an actor. And right now there is no way to create a delegate on actor creation. Although if we want to do that we should rethink more of this. Personally I think we shouldn't be special casing the actor channel as separate from the delegates.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 4]
```diff
@@ -932,30 +932,47 @@ void USpatialActorChannel::OnCreateEntityResponse(const Worker_CreateEntityRespo
...
-#if !UE_BUILD_SHIPPING
-		// Commands can timeout locally, but still be processed by the runtime. When this occurs, our follow up `CreateEntityRequest` will be
-		// superfluous, as the entity will already be created. If we detect that the entity is already in our view, reduce the message severity
-		else if (NetDriver->StaticComponentView->GetComponentData<SpatialGDK::Position>(GetEntityId()) == nullptr)
+		break;
+	case WORKER_STATUS_CODE_APPLICATION_ERROR:
+		if (bEntityIsInView)
 		{
-			UE_LOG(LogSpatialActorChannel, Error, TEXT("Failed to create entity for actor %s: Reason: %s"), *Actor->GetName(), UTF8_TO_TCHAR(Op.message));
+			UE_LOG(LogSpatialReceiver, Verbose, TEXT("Create entity request failed as the entity already exists and is in view. "
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: I feel this should be `Log` or even `Warning` as this seems to indicate the assumptions about how we deal with entity ID reservation don't hold.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 5]
```diff
@@ -932,30 +932,47 @@ void USpatialActorChannel::OnCreateEntityResponse(const Worker_CreateEntityRespo
...
+				"Actor %s, request id: %d, entity id: %lld, message: %s"), *Actor->GetName(), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
 		}
 		else
 		{
-			UE_LOG(LogSpatialActorChannel, Verbose, TEXT("Failed to create entity for actor %s, entity already in view : Reason: %s"), *Actor->GetName(), UTF8_TO_TCHAR(Op.message));
+			UE_LOG(LogSpatialReceiver, Log, TEXT("Create entity request failed."
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Why is this not a warning?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 6]
```diff
@@ -932,30 +932,47 @@ void USpatialActorChannel::OnCreateEntityResponse(const Worker_CreateEntityRespo
...
+				"Either the reservation expired, the entity already existed, or the entity was invalid. "
+				"Actor %s, request id: %d, entity id: %lld, message: %s"), *Actor->GetName(), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
 		}
-#endif // !UE_BUILD_SHIPPING
-
-		return;
+		break;
+	default:
+		UE_LOG(LogSpatialReceiver, Error, TEXT("Create entity request failed. This likely indicates a bug in the Unreal GDK and should be reported."
+				"Actor %s, request id: %d, entity id: %lld, message: %s"), *Actor->GetName(), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]:
```suggestion
			"Actor %s, request id: %d, entity id: %lld, message: %s"), *Actor->GetName(), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 7]
```diff
@@ -529,8 +529,8 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
 
 	if (AActor* EntityActor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId)))
 	{
-		UE_LOG(LogSpatialReceiver, Log, TEXT("Entity for actor %s has been checked out on the worker which spawned it or is a singleton linked on this worker"), \
-			*EntityActor->GetName());
+		UE_LOG(LogSpatialReceiver, Log, TEXT("Entity for actor %s has been checked out on the worker which spawned it or is a singleton linked on this worker. "
+		"Entity id: $lld"), *EntityActor->GetName(), EntityId);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
			"Entity id: $lld"), *EntityActor->GetName(), EntityId);
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 4 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: This will happen when the reservation was fine, but the create entity called timed-out locally but really succeeded. So you'd get this after retrying correctly and it turning out that stuff was fine all alone.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: In this case you can tell that happened as you can see it in the view.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 5 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Same as the comment above basically. It doesn't mean anything went wrong. It's kind of just an annoying part of retrying.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 6 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Is this specifically that you want it one tab back?

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Checked this one - agreed.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 5 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: I'd be fine with making this a warning as it's sort of an ephemeral failure (same as the above).
I guess I'm semi-anticipating that reservations expiring won't be a thing soon.
But yeah in the case where the entity was invalid or the reservation did expire that would be bad.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Hmm although maybe it could be argued that the timeout should be the warning then.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Discussed offline

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 3 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Discussed offline - will check actor channel is stale and remove this return. Also cleanup the delegate.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 4 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Discussed offline - will make it a log.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 8]
```diff
@@ -932,30 +932,47 @@ void USpatialActorChannel::OnCreateEntityResponse(const Worker_CreateEntityRespo
...
+	case WORKER_STATUS_CODE_TIMEOUT:
+		if (bEntityIsInView)
+		{
+			UE_LOG(LogSpatialActorChannel, Log, TEXT("Create entity request failed but the entity was already in view. "
+				"Actor %s, request id: %d, entity id: %lld, message: %s"), *Actor->GetName(), Op.request_id, Op.entity_id, UTF8_TO_TCHAR(Op.message));
+		}
+		else
 		{
-			UE_LOG(LogSpatialActorChannel, Warning, TEXT("Failed to create entity for actor %s Reason: %s. Retrying..."), *Actor->GetName(), UTF8_TO_TCHAR(Op.message));
+			UE_LOG(LogSpatialReceiver, Log, TEXT("Create entity request timed out. Retrying. "
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Should this perhaps be a warning

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 9]
```diff
@@ -1609,9 +1620,11 @@ void USpatialReceiver::OnCreateEntityResponse(const Worker_CreateEntityResponseO
 	{
 		Channel->OnCreateEntityResponse(Op);
 	}
-	else
+	else if (Channel.IsStale())
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: What happens if it's not valid and not stale?

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: The it shouldn't print any of these. This is here because we treat channels differently from other delegates. If you have neither (and never had a channel) then it just reverts to the verbose logging above.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 8 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Can leave it as a warning if we want.
